auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -1,25 +1,217 @@
+        Collection<IRole> roles = new ArrayList<IRole>();
+        if (userIdentity != null) {
+            Map<RoleKey, IRole> realmRoles = parseRealmRoles();
+            if (realmRoles != null) {
+                roles.addAll(realmRoles.values());
+            }
+            Map<RoleKey, IRole> clientRoles = parseClientRoles();
+            if (clientRoles != null) {
+                roles.addAll(clientRoles.values());
+            }
+        }
+        return roles;
+    }
+
+    /**
+     * Parses the user's realm roles from the user identity {@link #userIdentity}
+     *
+     * @return user's realm roles
+     * @throws JSONException
+     * @see RoleKey
+     * @see RealmRole
+     */
+    private Map<RoleKey, IRole> parseRealmRoles() throws JSONException {
+        Map<RoleKey, IRole> realmRoles = new HashMap<>();
+        if (userIdentity.has(REALM) && userIdentity.getJSONObject(REALM).has(ROLES)) {
+            String tokenRealmRolesJSON = userIdentity.getJSONObject(REALM).getString(ROLES);
+
+            String realmRolesString = tokenRealmRolesJSON.substring(1, tokenRealmRolesJSON.length() - 1).replace("\"", "");
+            String roles[] = realmRolesString.split(COMMA);
+
+            for (String rolename : roles) {
+                RealmRole realmRole = new RealmRole(rolename);
+                realmRoles.put(new RoleKey(realmRole, null), realmRole);
+            }
+        }
+        return realmRoles;
+    }
+
+    /**
+     * Parses the user's initial client roles from the user identity {@link #userIdentity}
+     *
+     * @return user's client roles
+     * @throws JSONException
+     * @see AuthServiceConfig for initial client
+     * @see RoleKey
+     * @see ClientRole
+     */
+    private Map<RoleKey, IRole> parseClientRoles() throws JSONException {
+        Map<RoleKey, IRole> clientRoles = new HashMap<>();
+
+        AuthServiceConfig authConfig = this.getConfig();
+        JSONObject authJSON =  authConfig.toJSON();
+
+        if (authJSON.has(RESOURCE)) {
+            String initialClientID = authJSON.get(RESOURCE).toString();  //immediate client role
+
+            if (userIdentity.has(CLIENT) && userIdentity.getJSONObject(CLIENT).has(initialClientID)
+                    && userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).has(ROLES)) {
+                String tokenClientRolesJSON = userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).getString(ROLES);
+
+                String clientRolesString = tokenClientRolesJSON.substring(1, tokenClientRolesJSON.length() - 1).replace("\"", "");
+                String roles[] = clientRolesString.split(COMMA);
+
+                for (String rolename : roles) {
+                    ClientRole clientRole = new ClientRole(rolename, initialClientID);
+                    clientRoles.put(new RoleKey(clientRole, initialClientID), clientRole);
+                }
+            }
+        }
+        return clientRoles;
```

Comment:
similarly, not sure why we return a map here. We can just return an array or list of ClientRole?